jmespath-core/src/main/java/io/burt/jmespath/function/CompareByFunction.java

```diff
@@ -0,0 +1,56 @@
+package io.burt.jmespath.function;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.ArrayList;
+
+import io.burt.jmespath.Adapter;
+import io.burt.jmespath.node.JmesPathNode;
+
+public class CompareByFunction extends JmesPathFunction {
+  private final int compareModifier;
+
+  public CompareByFunction(int compareModifier) {
+    super(2, 2);
+    this.compareModifier = compareModifier;
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    T array = arguments.get(0).value();
+    JmesPathNode expression = arguments.get(1).expression();
+    if (arguments.get(0).isExpression()) {
```

Comment:
I think you could solve it by providing a wider set of accessors for `ExpressionOrValue<T>`

That way, you could write something like:
```java
Iterator<T> elements = expressionOrValue.array(adapter).iterator();
```
where `array` could check that the object is an array (and return the underlying list), and raise an appropriate error type if the wrong type of element was returned.